File: go-backend-interview-prep/08_http_and_backend/11_sqlite_repo_pattern_intro/main.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"log"
	"net"
	"net/http"
	"strconv"
	"strings"
	"sync"
	"time"
)

// --- Domain ---

type Item struct {
	ID    int     `json:"id"`
	Name  string  `json:"name"`
	Price float64 `json:"price"`
}

// --- Repository interface ---
// In production, you'd have SQLiteRepo, PostgresRepo, etc.
// All implement this interface -- handlers never know which one.

type ItemRepo interface {
	Create(name string, price float64) (Item, error)
	GetByID(id int) (Item, error)
	List() ([]Item, error)
	Delete(id int) error
}

// --- In-memory implementation ---
// Go's stdlib has no SQLite driver. Real SQLite needs an external package.
// We demonstrate the pattern with a map.

type MemoryRepo struct {
	mu     sync.RWMutex
	items  map[int]Item
	nextID int
}

func NewMemoryRepo() *MemoryRepo {
	return &MemoryRepo{items: map[int]Item{}, nextID: 1}
}

var ErrNotFound = errors.New("item not found")

func (r *MemoryRepo) Create(name string, price float64) (Item, error) {
	r.mu.Lock()
	defer r.mu.Unlock()
	item := Item{ID: r.nextID, Name: name, Price: price}
	r.items[r.nextID] = item
	r.nextID++
	return item, nil
}

func (r *MemoryRepo) GetByID(id int) (Item, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	item, ok := r.items[id]
	if !ok {
		return Item{}, ErrNotFound
	}
	return item, nil
}

func (r *MemoryRepo) List() ([]Item, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	items := make([]Item, 0, len(r.items))
	for _, item := range r.items {
		items = append(items, item)
	}
	return items, nil
}

func (r *MemoryRepo) Delete(id int) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	if _, ok := r.items[id]; !ok {
		return ErrNotFound
	}
	delete(r.items, id)
	return nil
}

// --- Service layer (business logic) ---
// Thin in this demo, but in production this is where validation,
// business rules, and cross-cutting concerns live.

type ItemService struct {
	repo ItemRepo
}

func (s *ItemService) CreateItem(name string, price float64) (Item, error) {
	if name == "" {
		return Item{}, errors.New("name is required")
	}
	if price < 0 {
		return Item{}, errors.New("price must be non-negative")
	}
	return s.repo.Create(name, price)
}

func (s *ItemService) GetItem(id int) (Item, error) {
	return s.repo.GetByID(id)
}

func (s *ItemService) ListItems() ([]Item, error) {
	return s.repo.List()
}

func (s *ItemService) DeleteItem(id int) error {
	return s.repo.Delete(id)
}

// --- HTTP Handlers ---

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(v)
}

func makeHandlers(svc *ItemService) http.Handler {
	mux := http.NewServeMux()

	mux.HandleFunc("GET /items", func(w http.ResponseWriter, r *http.Request) {
		items, err := svc.ListItems()
		if err != nil {
			writeJSON(w, 500, map[string]string{"error": err.Error()})
			return
		}
		writeJSON(w, 200, items)
	})

	mux.HandleFunc("GET /items/{id}", func(w http.ResponseWriter, r *http.Request) {
		id, _ := strconv.Atoi(r.PathValue("id"))
		item, err := svc.GetItem(id)
		if errors.Is(err, ErrNotFound) {
			writeJSON(w, 404, map[string]string{"error": "not found"})
			return
		}
		if err != nil {
			writeJSON(w, 500, map[string]string{"error": err.Error()})
			return
		}
		writeJSON(w, 200, item)
	})

	mux.HandleFunc("POST /items", func(w http.ResponseWriter, r *http.Request) {
		var req struct {
			Name  string  `json:"name"`
			Price float64 `json:"price"`
		}
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			writeJSON(w, 400, map[string]string{"error": "invalid json"})
			return
		}
		item, err := svc.CreateItem(req.Name, req.Price)
		if err != nil {
			writeJSON(w, 400, map[string]string{"error": err.Error()})
			return
		}
		writeJSON(w, 201, item)
	})

	mux.HandleFunc("DELETE /items/{id}", func(w http.ResponseWriter, r *http.Request) {
		id, _ := strconv.Atoi(r.PathValue("id"))
		err := svc.DeleteItem(id)
		if errors.Is(err, ErrNotFound) {
			writeJSON(w, 404, map[string]string{"error": "not found"})
			return
		}
		writeJSON(w, 200, map[string]string{"deleted": strconv.Itoa(id)})
	})

	return mux
}

func main() {
	// Wire up: repo -> service -> handlers
	repo := NewMemoryRepo()
	svc := &ItemService{repo: repo}
	handler := makeHandlers(svc)

	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		log.Fatal(err)
	}
	addr := ln.Addr().String()
	fmt.Printf("Repo Pattern API on http://%s\n", addr)
	fmt.Println("  Architecture: handler -> service -> repository (interface)")
	fmt.Println("  Storage: in-memory map (swap to SQLite/Postgres by changing repo)")

	server := &http.Server{Handler: handler}

	// Self-test
	go func() {
		time.Sleep(100 * time.Millisecond)
		base := "http://" + addr
		fmt.Println("\n--- Demo ---")

		// Create
		resp, _ := http.Post(base+"/items", "application/json",
			strings.NewReader(`{"name":"widget","price":9.99}`))
		if resp != nil {
			var item Item
			json.NewDecoder(resp.Body).Decode(&item)
			resp.Body.Close()
			fmt.Printf("  CREATE: %+v\n", item)
		}

		http.Post(base+"/items", "application/json",
			strings.NewReader(`{"name":"gadget","price":19.99}`))

		// List
		resp2, _ := http.Get(base + "/items")
		if resp2 != nil {
			var items []Item
			json.NewDecoder(resp2.Body).Decode(&items)
			resp2.Body.Close()
			fmt.Printf("  LIST:   %d items\n", len(items))
		}

		// Get
		resp3, _ := http.Get(base + "/items/1")
		if resp3 != nil {
			var item Item
			json.NewDecoder(resp3.Body).Decode(&item)
			resp3.Body.Close()
			fmt.Printf("  GET:    %+v\n", item)
		}

		// Delete
		delReq, _ := http.NewRequest("DELETE", base+"/items/2", nil)
		resp4, _ := http.DefaultClient.Do(delReq)
		if resp4 != nil {
			var result map[string]string
			json.NewDecoder(resp4.Body).Decode(&result)
			resp4.Body.Close()
			fmt.Printf("  DELETE: %v\n", result)
		}

		// 404
		resp5, _ := http.Get(base + "/items/99")
		if resp5 != nil {
			resp5.Body.Close()
			fmt.Printf("  GET 99: status %d\n", resp5.StatusCode)
		}

		fmt.Println("\nDemo done.")
		server.Close()
	}()

	if err := server.Serve(ln); err != http.ErrServerClosed {
		log.Fatal(err)
	}
}

File: go-backend-interview-prep/08_http_and_backend/11_sqlite_repo_pattern_intro/main.py
"""SQLite / Repository pattern intro -- Python equivalent of the Go example.

Python has sqlite3 in the stdlib, so we use a real SQLite database (in-memory).
Demonstrates the layered architecture: handler -> service -> repository.
"""

import json
import sqlite3
import threading
import urllib.request
import urllib.error
from http.server import HTTPServer, BaseHTTPRequestHandler


# --- Repository interface (duck typing in Python) ---

class SQLiteRepo:
    """Repository backed by sqlite3 (stdlib). Real database, not a dict."""

    def __init__(self, db_path: str = ":memory:"):
        self.conn = sqlite3.connect(db_path, check_same_thread=False)
        self.conn.row_factory = sqlite3.Row
        self.lock = threading.Lock()
        self.conn.execute("""
            CREATE TABLE IF NOT EXISTS items (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                name TEXT NOT NULL,
                price REAL NOT NULL
            )
        """)
        self.conn.commit()

    def create(self, name: str, price: float) -> dict:
        with self.lock:
            cur = self.conn.execute(
                "INSERT INTO items (name, price) VALUES (?, ?)",
                (name, price),
            )
            self.conn.commit()
            return {"id": cur.lastrowid, "name": name, "price": price}

    def get_by_id(self, item_id: int) -> dict | None:
        with self.lock:
            row = self.conn.execute(
                "SELECT id, name, price FROM items WHERE id = ?", (item_id,)
            ).fetchone()
        if row is None:
            return None
        return {"id": row["id"], "name": row["name"], "price": row["price"]}

    def list_all(self) -> list[dict]:
        with self.lock:
            rows = self.conn.execute("SELECT id, name, price FROM items").fetchall()
        return [{"id": r["id"], "name": r["name"], "price": r["price"]} for r in rows]

    def delete(self, item_id: int) -> bool:
        with self.lock:
            cur = self.conn.execute("DELETE FROM items WHERE id = ?", (item_id,))
            self.conn.commit()
            return cur.rowcount > 0

    def close(self):
        self.conn.close()


# --- Service layer ---

class ItemService:
    def __init__(self, repo):
        self.repo = repo

    def create_item(self, name: str, price: float) -> dict:
        if not name:
            raise ValueError("name is required")
        if price < 0:
            raise ValueError("price must be non-negative")
        return self.repo.create(name, price)

    def get_item(self, item_id: int) -> dict | None:
        return self.repo.get_by_id(item_id)

    def list_items(self) -> list[dict]:
        return self.repo.list_all()

    def delete_item(self, item_id: int) -> bool:
        return self.repo.delete(item_id)


# --- HTTP Handlers ---

# Module-level service (set in main)
svc: ItemService | None = None


def json_response(handler, status: int, data) -> None:
    handler.send_response(status)
    handler.send_header("Content-Type", "application/json")
    handler.end_headers()
    handler.wfile.write(json.dumps(data).encode())


def read_json(handler) -> dict:
    length = int(handler.headers.get("Content-Length", 0))
    return json.loads(handler.rfile.read(length)) if length else {}


class Handler(BaseHTTPRequestHandler):
    def do_GET(self):
        import re

        if self.path == "/items":
            items = svc.list_items()
            json_response(self, 200, items)
            return

        m = re.match(r"^/items/(\d+)$", self.path)
        if m:
            item = svc.get_item(int(m.group(1)))
            if item is None:
                json_response(self, 404, {"error": "not found"})
                return
            json_response(self, 200, item)
            return

        json_response(self, 404, {"error": "not found"})

    def do_POST(self):
        if self.path == "/items":
            body = read_json(self)
            try:
                item = svc.create_item(body.get("name", ""), body.get("price", 0))
                json_response(self, 201, item)
            except ValueError as e:
                json_response(self, 400, {"error": str(e)})
            return
        json_response(self, 404, {"error": "not found"})

    def do_DELETE(self):
        import re
        m = re.match(r"^/items/(\d+)$", self.path)
        if m:
            deleted = svc.delete_item(int(m.group(1)))
            if not deleted:
                json_response(self, 404, {"error": "not found"})
                return
            json_response(self, 200, {"deleted": m.group(1)})
            return
        json_response(self, 404, {"error": "not found"})

    def log_message(self, *args):
        pass


def main() -> None:
    global svc

    # Wire up: repo (SQLite) -> service -> handlers
    repo = SQLiteRepo(":memory:")
    svc = ItemService(repo)

    server = HTTPServer(("127.0.0.1", 0), Handler)
    host, port = server.server_address
    base = f"http://{host}:{port}"
    print(f"Repo Pattern API on {base}")
    print("  Architecture: handler -> service -> repository (SQLite)")
    print("  Storage: sqlite3 in-memory (stdlib)")
    threading.Thread(target=server.serve_forever, daemon=True).start()

    print("\n--- Demo ---")

    # Create
    req = urllib.request.Request(
        f"{base}/items",
        data=json.dumps({"name": "widget", "price": 9.99}).encode(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    resp = urllib.request.urlopen(req)
    print(f"  CREATE: {resp.read().decode().strip()}")

    # Create another
    req2 = urllib.request.Request(
        f"{base}/items",
        data=json.dumps({"name": "gadget", "price": 19.99}).encode(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    urllib.request.urlopen(req2).read()

    # List
    resp3 = urllib.request.urlopen(f"{base}/items")
    items = json.loads(resp3.read())
    print(f"  LIST:   {len(items)} items")

    # Get
    resp4 = urllib.request.urlopen(f"{base}/items/1")
    print(f"  GET:    {resp4.read().decode().strip()}")

    # Delete
    req5 = urllib.request.Request(f"{base}/items/2", method="DELETE")
    resp5 = urllib.request.urlopen(req5)
    print(f"  DELETE: {resp5.read().decode().strip()}")

    # 404
    try:
        urllib.request.urlopen(f"{base}/items/99")
    except urllib.error.HTTPError as e:
        print(f"  GET 99: status {e.code}")

    server.shutdown()
    repo.close()
    print("\nDone.")


if __name__ == "__main__":
    main()
